rllib/agents/trainer_config.py

```diff
@@ -0,0 +1,1015 @@
+        num_cpus_for_local_worker: Optional[int] = None,
+        custom_resources_per_worker: Optional[dict] = None,
+        placement_strategy: Optional[str] = None,
+    ) -> "TrainerConfig":
+        """Specifies resources allocated for a Trainer and its ray actors/workers.
+
+        Args:
+            num_gpus: Number of GPUs to allocate to the trainer process.
+                Note that not all algorithms can take advantage of trainer GPUs.
+                Support for multi-GPU is currently only available for
+                tf-[PPO/IMPALA/DQN/PG]. This can be fractional (e.g., 0.3 GPUs).
+            _fake_gpus: Set to True for debugging (multi-)?GPU funcitonality on a
+                CPU machine. GPU towers will be simulated by graphs located on
+                CPUs in this case. Use `num_gpus` to test for different numbers of
+                fake GPUs.
+            num_cpus_per_worker: Number of CPUs to allocate per worker.
+            num_gpus_per_worker: Number of GPUs to allocate per worker. This can be
+                fractional. This is usually needed only if your env itself requires a
+                GPU (i.e., it is a GPU-intensive video game), or model inference is
+                unusually expensive.
+            custom_resources_per_worker: Any custom Ray resources to allocate per
+                worker.
+            num_cpus_for_local_worker: Number of CPUs to allocate for the trainer.
+                Note: this only takes effect when running in Tune. Otherwise,
+                the trainer runs in the main program (driver).
+            custom_resources_per_worker: Any custom Ray resources to allocate per
+                worker.
+            placement_strategy: The strategy for the placement group factory returned by
+                `Trainer.default_resource_request()`. A PlacementGroup defines, which
+                devices (resources) should always be co-located on the same node.
+                For example, a Trainer with 2 rollout workers, running with
+                num_gpus=1 will request a placement group with the bundles:
+                [{"gpu": 1, "cpu": 1}, {"cpu": 1}, {"cpu": 1}], where the first bundle
+                is for the driver and the other 2 bundles are for the two workers.
+                These bundles can now be "placed" on the same or different
+                nodes depending on the value of `placement_strategy`:
+                "PACK": Packs bundles into as few nodes as possible.
+                "SPREAD": Places bundles across distinct nodes as even as possible.
+                "STRICT_PACK": Packs bundles into one node. The group is not allowed
+                    to span multiple nodes.
+                "STRICT_SPREAD": Packs bundles across distinct nodes.
+
+        Returns:
+            This updated TrainerConfig object.
+        """
+        if num_gpus is not None:
+            self.num_gpus = num_gpus
```

Comment:
maybe have a helper method to simplify this?

`pip install -U varname` 


```python

def set(self, value):
    from varname import nameof
    if value is not None:
        setattr(self, nameof(value), value)
```

and then in code use this as:

```python
self.set(num_gpus)
```

Reply:
but only if we feel it's not too much magic for such a basis operation.

Reply:
Yeah, would shorten things a bit, but I feel like it's ok as we have it now. It's mostly work in only the TrainerConfig dict, not in the sub-classes (like PPOConfig).

Reply:
oh, that's true. never mind, I thought we'd have this in all subclasses.